[GPU] Enable Metal on macosx

xta0 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48008 [Person Seg] Compress the person seg model
* **#47635 [GPU] Enable Metal on macosx**

Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

xta0 committed:
[GPU] Enable Metal on macosx

Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

ljk53 commented on cmake/Metal.cmake:
```diff
@@ -0,0 +1,42 @@
+exec_program(/usr/bin/xcode-select ARGS -print-path OUTPUT_VARIABLE CMAKE_XCODE_DEVELOPER_DIR)
+set(XCODE_POST_43_ROOT "${CMAKE_XCODE_DEVELOPER_DIR}/Platforms/${OSX_PLATFORM}/Developer")
+set(XCODE_PRE_43_ROOT "/Developer/Platforms/${OSX_PLATFORM}/Developer")
+if(NOT DEFINED CMAKE_OSX_DEVELOPER_ROOT)
+    if(EXISTS ${XCODE_POST_43_ROOT})
+        set(CMAKE_OSX_DEVELOPER_ROOT ${XCODE_POST_43_ROOT})
+    elseif(EXISTS ${XCODE_PRE_43_ROOT})
+        set(CMAKE_OSX_DEVELOPER_ROOT ${XCODE_PRE_43_ROOT})
+    endif(EXISTS ${XCODE_POST_43_ROOT})
+endif(NOT DEFINED CMAKE_OSX_DEVELOPER_ROOT)
+set(CMAKE_OSX_DEVELOPER_ROOT ${CMAKE_OSX_DEVELOPER_ROOT} CACHE PATH "Location of OSX SDKs root directory")
+
+if(NOT DEFINED CMAKE_OSX_SDK_ROOT)
+    file(GLOB _CMAKE_OSX_SDKS "${CMAKE_OSX_DEVELOPER_ROOT}/SDKs/*")
+    if(_CMAKE_OSX_SDKS) 
+        list(SORT _CMAKE_OSX_SDKS)
+        list(REVERSE _CMAKE_OSX_SDKS)
+        list(GET _CMAKE_OSX_SDKS 0 CMAKE_OSX_SDK_ROOT)
+        message(STATUS "_CMAKE_OSX_SDKS: ${_CMAKE_OSX_SDKS}")
+    else(_CMAKE_OSX_SDKS)
+        message(FATAL_ERROR "No OSX SDK's found in default search path ${CMAKE_OSX_DEVELOPER_ROOT}.")
+    endif(_CMAKE_OSX_SDKS)
+    message(STATUS "Toolchain using default OSX SDK: ${CMAKE_OSX_SDK_ROOT}")
+endif(NOT DEFINED CMAKE_OSX_SDK_ROOT)
```
Seems very similar to iOS.cmake - any chance to factor out?
also cc: @malfet @xuhdev

ljk53 commented on aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:
```diff
@@ -3,7 +3,7 @@
 
 
-#if defined(C10_IOS)
+#if (C10_IOS || TARGET_OS_MAC)
```
Nothing wrong about this - just feel that having C10_IOS (defined by us) and TARGET_OS_MAC (defined by apple) on the same condition is a little unbalanced, lol.

If this condition is going to be used at other places, shall we create a dedicated macro? Not sure what's the best convention in PyTorch codebase (cc: @dzhulgakov):
- if it's targeting apple devices, shall we introduce 'C10_APPLE'?
- or, can we simply use some master feature switch? e.g. '#if AT_METAL_ENABLED()' seems to be one of the common pattern.

xta0 replied:
How about (TARGET_OS_PHONE || TARGET_OS_MAC). Would that make you feel better 🤣

ljk53 commented on aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:
```diff
@@ -94,19 +94,19 @@ c10::intrusive_ptr<Conv2dOpContext> conv2d_prepack(
 #else
-  TORCH_CHECK(false, "conv2d_prepack_run can only be invoked on iOS");
+  TORCH_CHECK(false, "conv2d_prepack_run can only be invoked on iOS and MacOS");
   return input;
```
nit: I don't think we need 'return input;' here as TORCH_CHECK(false, ...) throws.

ljk53 commented on aten/src/ATen/native/metal/MetalPrepackOpRegister.cpp:
```diff
@@ -94,19 +94,19 @@ c10::intrusive_ptr<Conv2dOpContext> conv2d_prepack(
 #else
-  TORCH_CHECK(false, "copy_to_host can only be invoked on iOS");
+  TORCH_CHECK(false, "copy_to_host can only be invoked on iOS and MacOS");
   return input;
```
ditto

ljk53 commented on caffe2/CMakeLists.txt:
```diff
@@ -1195,6 +1195,18 @@ if(USE_CUDA)
 endif()
 
+# ---[ Metal(OSX) modification
+if(APPLE AND USE_PYTORCH_METAL)
```
nit: can we make 'USE_PYTORCH_METAL' only enabled when 'APPLE' is true?

xta0 replied:
`USE_PYTORCH_METAL` is also available in Linux. In that case, only one cpp file under `aten/native/metal` will be compiled. The reason is that users can still do `optimize_for_mobile` on Linux machines to export models for metal.

ljk53 reviewed (commented):
LGTM, just some nits and questions about convention.

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

ljk53 reviewed (approved):
accept to unblock - we could clean up these bool conditions (C10_IOS || TARGET_OS_MAC) later if they are going to be used a lot in the codebase in the future.

xta0 committed:
Update on "[GPU] Enable Metal on macosx"


Add macosx support for metal. The supported os version is 10.13 and above.

Differential Revision: [D24825088](https://our.internmc.facebook.com/intern/diff/D24825088/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D24825088/)!

[ghstack-poisoned]

PeterL1n commented:
Does this mean PyTorch will support inferencing and training on Mac using GPU in the future?

leovinus2001 commented:
> Does this mean PyTorch will support inferencing and training on Mac using GPU in the future?

I am wondering the same. Could someone in the know write a few sentences pls as to the goal here? 
Train vs inference?
macOS vs iOS?
Metal throughout PyTorch or just ATen (base don checkins I see :)
Metal just for CNN or also Linear/LSTM?

When more clarity is provided then it is easier to see whether we can contribute :)
